### DISCUSSION TITLE
setting up for windows users

### DISCUSSION THREAD

**Discussion by midicdj1000** [picocomputer/community#27] [Q&A]
I am new to the picocomputer and always work on windows so I have tried to write a getting started to allow the ehbasic to be compiled

`
install python by going to a command prompt and typing python
if it's not installed it will open the windows store and then install python 3.11
run python3.11 and check if pip is installed 

python3.11 pip3 -v

if not you can download from the windows command promt 

curl https://bootstrap.pypa.io/get-pip.py -o get-pip.py

then install with 
python3.11 get-pip.py

this will then let you install pyserial
python3.11 -m pip install pyserial

Download 64tass from
https://sourceforge.net/projects/tass64/files/latest/download
extract this folder and store on the project drive. I am using d:\64tass

download cc65 from
https://sourceforge.net/projects/cc65/files/cc65-snapshot-win32.zip
extract this to d:\cc65

these need to be added to the path in windows
open control panel, system , advanced system settings
click environmental variables.
select path and edit
select new and it should allow editing on a new line
type the location of 64tass so for me d:\64tass
select new and add d:\cc65

click ok and ok

close any command prompts so the new path is updated

New we can work to compile ehbasic and upload to the pico6502
Download the repository from
https://github.com/picocomputer/ehbasic
just like before I am going to make a project folder on my D drive with easy to find names
d:\pico6502
then extract the ehbasic inside this
so you should have the folder D:\pico6502\ehbasic-main

go to a command prompt and type 
cd D:\pico6502\ehbasic-main

the github instructions say 
See basic.py so not very helpful

I like to use notepad++ to edit files and we need to open basic.py to make some changes to the serial port commands because it’s setup for linux

Make sure the picocomputer is connected and you have a usb serial device and note the com port. In my case it was com3.
In basic.py it has two lines 
177 and 209
Mon=Monitor(‘/dev/attyACM0’)
Change this to Mon=Monitor(‘com3’) or what ever your com port is.
Then at the very bottom is has the command to make the basic rom
print("Usage: python3 basic.py run|upload")

don’t just copy and paste this instead from your command prompt that should be in the D:\pico6502\ehbasic-main folder type
python3.11 basic.py upload
this should then compile the code and upload it to the connected picocomputer
`

**Comment by brentward** [comment]
Maybe you could write this up as a pull request (PR) for the ehbasic repository so it will be included in the instructions on github. Help has been requested with writing documentation, although it is up to the maintainer what gets merged.

If you are unfamiliar with the process of creating a PR for a github repository and don't want to learn git in the command line the process for creating a PR directly in the browser is documented in github's documentation.

**Reply by midicdj1000** [reply]
have never written anything like a PR so I will have to take a look at the documentation.

I do have a video that is about to be set public when I build the kit and get VScode working under windows

https://www.youtube.com/watch?v=zJpz16XDL9c

**Comment by midicdj1000** [comment]
Ok, I set my video to public so this is it, I build the kit and then go through all the steps to get the hello.c to compile in windows vscode.

https://youtu.be/zJpz16XDL9c

I have not setup the usb/ip part yet.

**Comment by WojciechGw** [comment]
After a long fight with setting my programming environment for Picocomputer 6502 ...
I want to share my journal about it.
This is of course my case, but maybe it will save you a lot of stress and time.

**! Please remember that I am not responsible for your actions and their effects based on the description below.**

target :
Picocomputer 6502 developer environment configuration:
Microsoft Windows 10 22H2 19045.3693 (untested under Windows 11)
with Microsoft VSCode (updated to newest)
with Microsoft WSL2 with Ubuntu (upgraded to newest)
with VSCode Remote Connetion to WSL2/Ubuntu

IMHO don't loose Your time for virtual setups like VirtualBox Ubuntu/Debian/Windows or whatever

Windows PowerShell
==================
`PS C:>wsl --update`

then connect Picocomputer 6502 and/or Raspberry Picoprobe(s) (USB)

install usbipd-win from
https://github.com/dorssel/usbipd-win/releases/tag/v3.2.0

then

`PS C:\>usbipd wsl list`

```
BUSID  VID:PID    DEVICE                                                        STATE
3-2    2e8a:000c  CMSIS-DAP v2 Interface, USB serial device (COM9)              Not attached
3-3    2e8a:000c  CMSIS-DAP v2 Interface, USB serial device (COM8)              Not attached
3-4    2e8a:4001  USB serial device (COM6)                                      Not attached
```

for 1st time as administrator !
the order matters for WSL2/Ubuntu's ttyACMx assignment

/dev/ttyACM0 Picocomputer 6502 for minicom and VScode connections
`PS C:\>usbipd wsl attach --busid 3-4`

/dev/ttyACM1 picoprobe #1 RIA (in my case)
`PS C:\>usbipd wsl attach --busid 3-3`

/dev/ttyACM2 picoprobe #2 VGA (in my case)
`PS C:\>usbipd wsl attach --busid 3-2`

after that

`PS C:\> usbipd wsl list`
```
BUSID  VID:PID    DEVICE                                                        STATE
3-2    2e8a:000c  CMSIS-DAP v2 Interface, USB serial device (COM9)              Attached - WSL
3-3    2e8a:000c  CMSIS-DAP v2 Interface, USB serial device (COM8)              Attached - WSL
3-4    2e8a:4001  USB serial device (COM6)                                      Attached - WSL
```

! unfortunately You will have to do this every time before starting VSCode 
! or when ttyACMx disappears from WSL2/Ubuntu
! maybe some PS scripts for automation ?

WSL2 Ubuntu
===========

1st
============
/etc/wsl.conf
edit or create
-------------------------------
```
[boot]
systemd=true
command=service udev start
```

2nd
============
add at the end of file
~/.bashrc
-------------------------------
```
# start udev
sudo /usr/sbin/service udev start
```

3rd (optional)
================================
`$sudo visudo`
and add at the end of file
-------------------------------
`<your_username> ALL=(ALL) NOPASSWD: /usr/sbin/service udev start`

4th
============
/etc/udev/rules.d/99-pico.rules
edit or create
-------------------------------
```
# Make RP2040 in BOOTSEL mode writable by all users
# without sudo nightmare
SUBSYSTEM=="usb", ATTRS{idVendor}=="2e8a", ATTRS{idProduct}=="0003", MODE="0666"
SUBSYSTEM=="usb", ATTRS{idVendor}=="2e8a", ATTRS{idProduct}=="4001", MODE="0666"
SUBSYSTEM=="usb", ATTRS{idVendor}=="2e8a", ATTRS{idProduct}=="000c", MODE="0666"

# Pico for python
SUBSYSTEM=="tty", ATTRS{idVendor}=="2e8a", ATTRS{idProduct)=="0005", SYMLINK+="pico"
SUBSYSTEM=="tty", ATTRS{idVendor}=="2e8a", ATTRS{idProduct)=="4001", SYMLINK+="pico"
SUBSYSTEM=="tty", ATTRS{idVendor}=="2e8a", ATTRS{idProduct)=="000c", SYMLINK+="pico"

# Picoprobe
SUBSYSTEM=="usb", ATTRS{idVendor}=="2e8a", MODE="0666"

# after editing
# sudo udevadm control --reload-rules && sudo udevadm trigger
# connect and disconnect the USB device
```

finally
============
at WSL2/Ubuntu test connection
------------------------------

```
$cd ~
$mkdir <your_dirname>
$cd <your_dirname>
$git clone https://github.com/picocomputer/cc65.git
$cd cc65
$make all
$sudo make install
$sudo make clean
$cd ..
```
create in your github repository <your_name_for_repository> from Rumbledethups' template rp6502-vscode
`$git clone https://github.com/<your_github>/<your_name_for_repository>.git
`
```
$git clone https://github.com/picocomputer/examples.git
$git clone https://github.com/picocomputer/ehbasic.git
$git clone https://github.com/picocomputer/adventure.git
$cd adventure
$git submodule update --init
$cd ..
```

optionaly for Picocomputer 6502 firmware developing
`$git clone https://github.com/picocomputer/rp6502.git`

`$ minicom -c on -b 115200 -D /dev/ttyACM0`
test some command
`]help`
leave minicom

Windows VSCode 
--------------
establish remote connection to WSL
open folder (from WSL/Ubuntu) with desired repository
Build what You want
Run Build Task ... (Shift+Ctrl+B)

:) mission accomplished, have fun :)
